From 59d58b1e9ca2b48e3e2cf65004041257c40465aa Mon Sep 17 00:00:00 2001
From: "Ooi, Joyce" <joyce.ooi@intel.com>
Date: Mon, 16 Mar 2020 12:35:11 +0800
Subject: [PATCH 087/120] net: ethernet: altera: fix TOD adjust time

commit c0f1b68289990487306d2d4ac5d562908446b55e from
https://github.com/altera-opensource/linux-socfpga.git

This patch fixes the TOD adjust time by adding a check on the delta
value before performing the coarse time offset adjustment. This prevents
a negative delta value to cause the master offset to be calculated
incorrectly.

Signed-off-by: Ooi, Joyce <joyce.ooi@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/altera/intel_fpga_tod.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/ethernet/altera/intel_fpga_tod.c b/drivers/net/ethernet/altera/intel_fpga_tod.c
index e3af41077057..a40fde583d30 100644
--- a/drivers/net/ethernet/altera/intel_fpga_tod.c
+++ b/drivers/net/ethernet/altera/intel_fpga_tod.c
@@ -190,6 +190,9 @@ static int intel_fpga_tod_adjust_time(struct ptp_clock_info *ptp, s64 delta)
 	 * just set the time.
 	 */
 	if (count > TOD_ADJUST_COUNT_MAX) {
+		if (neg_adj)
+			delta = -delta;
+
 		/* Perform the coarse time offset adjustment */
 		ret = coarse_adjust_tod_clock(priv, delta);
 	} else {
-- 
2.31.1

